prime-router/src/main/kotlin/azure/ReportFunction.kt

```diff
@@ -594,31 +581,68 @@ class ReportFunction : Logging {
                 return sb.toString()
             }
 
-            fun writeConsolidatedArray(field: String, array: List<ResultDetail>) {
-                val rowsByGroupingId = hashMapOf<String, MutableList<Int>>()
+            fun createPropertiesByGroupingId(array: List<ResultDetail>): List<HashMap<String, out Any>> {
+                val itemsByGroupingId = hashMapOf<String, MutableList<Int>>()
                 val messageByGroupingId = hashMapOf<String, String>()
+                val scopesByGroupingId = hashMapOf<String, String>()
+                val itemDetailsWithGroupingId = hashMapOf<String, String>()
                 array.forEach { resultDetail ->
                     val groupingId = resultDetail.responseMessage.groupingId()
-                    if (!rowsByGroupingId.containsKey(groupingId)) {
-                        rowsByGroupingId[groupingId] = mutableListOf()
+                    if (!itemsByGroupingId.containsKey(groupingId)) {
+                        itemsByGroupingId[groupingId] = mutableListOf()
                         messageByGroupingId[groupingId] = resultDetail.responseMessage.detailMsg()
+                        scopesByGroupingId[groupingId] = resultDetail.scope.toString()
                     }
                     if (resultDetail.row != -1) {
                         // Add 2 to account for array offset and csv header
-                        rowsByGroupingId[groupingId]?.add(resultDetail.row + 2)
+                        itemsByGroupingId[groupingId]?.add(resultDetail.row + 2)
+                        itemDetailsWithGroupingId[(resultDetail.row + 2).toString()] =
+                            resultDetail.responseMessage.groupingId()
                     }
                 }
+                return listOf<HashMap<String, out Any>>(
```

Comment:
Per my comment above, change to construct a data class obj that acts as a container for these 4 Maps - easier to read.